org-code-javabuilder/lib/src/main/java/org/code/javabuilder/LambdaRequestHandler.java

```diff
@@ -42,30 +40,45 @@
   // This is an error code we made up to signal that available disk space is less than 50%.
   // It may be used in tracking on Cloud Watch.
   private static final int LOW_DISK_SPACE_ERROR_CODE = 50;
+  private static final String CONTENT_BUCKET_NAME = System.getenv("CONTENT_BUCKET_NAME");
+  private static final String CONTENT_BUCKET_URL = System.getenv("CONTENT_BUCKET_URL");
+  private static final String API_ENDPOINT = System.getenv("API_ENDPOINT");
+  private final AmazonApiGatewayManagementApi API_CLIENT;
+  private final AmazonSQS SQS_CLIENT;
+  private final AmazonS3 S3_CLIENT;
 
   public LambdaRequestHandler() {
     // create CachedResources once for the entire container.
     // This will only be called once in the initial creation of the lambda instance.
     // Documentation: https://docs.aws.amazon.com/lambda/latest/dg/java-handler.html
     CachedResources.create();
+
+    // Creating these clients here rather than in the request handler method allows us to use
+    // provisioned concurrency to decrease cold boot time by 3-10 seconds, depending on the lambda
```

Comment:
clarifying question: my gut feeling on "cold boot" is it would be running the init stage, so we would want less logic in the constructor. Clearly that gut feeling is wrong, can you explain why putting more in the constructor makes cold boot faster? Is there some prior stage to cold boot that I'm missing?

Reply:
There's an `Init` phase which "downloads the code for the function and all layers, initializes any extensions, initializes the runtime, and then runs the function’s initialization code (the code outside the main handler)." For lambdas that have provisioned concurrency set, this all happens before the first function invocation, so long as it takes less than 10s. 

Details there came from this doc: [Lambda execution lifecycle](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-context.html). 

There was also a note on the [java-specific docs](https://docs.aws.amazon.com/lambda/latest/dg/java-handler.html) that seems to indicate that the class constructor also runs during the init phase. "You can add initialization code outside of your handler method to reuse resources across multiple invocations. When the runtime loads your handler, it runs static code and the class constructor. Resources that are created during initialization stay in memory between invocations, and can be reused by the handler thousands of times."

Reply:
ah makes sense. My confusion was I thought our issue was being over that 10 second limit already, when really we weren't utilizing that limit to its fullest.

Reply:
Oh yeah, good callout. I forgot to clarify that once I discovered that was the case.